### PR TITLE
plugins/startup: init

### LINKED ISSUE #791: startup.nvim
**Plugin**: startup.nvim
**Plugin repository**: [URL to the repository of the plugin](https://github.com/startup-nvim/startup.nvim)
**Is the plugin on nixpkgs**: No

## Extra information
Customizable startup screen for neovim, similar to startify. 
Depends on plenary.nvim, and should depend on telescope.nvim for default setup.
I don't really understand fully how nixvim works, so I'm not sure how to add this myself, but if it would be simple for someone to implement this, I'd be grateful!

### PR DESCRIPTION
**Pull request by GaetanLepage**
Add support for [startup.nvim](https://github.com/startup-nvim/startup.nvim).

Fixes #791

### CODE REVIEW

File: plugins/utils/startup.nix [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+      {
+        assertion = (cfg.options.paddings == null) || (length cfg.options.paddings) == numSections;
+        message = ''
+          Nixvim (plugins.startup): Make sure that `plugins.startup.options.paddings` has the same
+          number of elements as there are sections.
+        '';
+      }
+      {
+        assertion =
+          ((length cfg.parts) == numSections)
```
Comment (traxys) [inline plugins/utils/startup.nix:276]: I think this should be a `<` you could define more sections than you use in parts

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/startup.nix (hunk 1 above)
Comment (GaetanLepage) [reply plugins/utils/startup.nix:276]: I don't think so. In the README, it is mentioned "all sections in order" for the `parts` options.

Comment (traxys) [reply plugins/utils/startup.nix:276]: I think this means "in order you want them to be shown"

Comment (traxys) [reply plugins/utils/startup.nix:276]: Lookig at the code it seems that way:
```
    for _, part in ipairs(parts) do
        utils.empty(settings.options.paddings[padding_nr])
        padding_nr = padding_nr + 1
        current_section = part
        local options = settings[part]
        options = utils.validate_settings(options)
```

Comment (GaetanLepage) [reply plugins/utils/startup.nix:276]: Ok, I switched to `<` and changed the assertion message.

File: plugins/utils/startup.nix [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+          number of elements as there are sections.
+        '';
+      }
+      {
+        assertion =
+          ((length cfg.parts) <= numSections)
+          && (
+            all
+            (sectionName: elem sectionName cfg.parts)
+            sectionNames
```
Comment (traxys) [inline plugins/utils/startup.nix:280]: Sorry this one seems off too, I think you need to check `all (part: builtins.hasAttr part cfg.sections) cfg.parts` in order to check that all parts are defined.

And the message should be `You should not have parts that are not defined in 'plugins.startup.sections'`, I think that condition also avoids the need to check the length

**Review by traxys — CHANGES_REQUESTED** [review]

File: plugins/utils/startup.nix (hunk 2 above)
Comment (GaetanLepage) [reply plugins/utils/startup.nix]: Done !

**Review by traxys — APPROVED** [review]

**Pull request merged**